### PR TITLE
transport: sprinkle trace logs around the code

### PR DESCRIPTION
**Pull request by psarna**
In order to improve debugging experience, more traces are added.

Also, the error message informing that there are no connections in the pool has a little more context - it distinguishes between broken and initializing state.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

File: scylla/src/transport/connection_pool.rs [hunk 1]
```diff
@@ -80,6 +90,43 @@ enum PoolConnections {
...
+        match self {
+            PoolConnections::NotSharded(conns) => {
+                write!(f, "non-sharded: {}", format_connection_vector(conns))
+            }
+            PoolConnections::Sharded {
+                sharder,
+                connections,
+            } => write!(
+                f,
+                "sharded({};{}): {}",
```
Comment (piodul) [inline scylla/src/transport/connection_pool.rs:121]: The message format looks a bit cryptic to the outsider (aka. user). How about this:

```
sharded(nr_shards: {}, msb_ignore_bits: {}): ...
```

The `...` part, for which I was too lazy to come up with a precise format string, we could include information about how many connections are there for each shard.

File: scylla/src/transport/connection_pool.rs [hunk 2]
```diff
@@ -80,6 +90,43 @@ enum PoolConnections {
     },
 }
 
+fn format_connection_vector(conns: &[Arc<Connection>]) -> String {
+    conns
+        .iter()
+        .map(|conn| conn.get_connect_address().to_string())
+        .collect::<Vec<String>>()
+        .join(",")
```
Comment (piodul) [inline scylla/src/transport/connection_pool.rs:98]: Won't the addresses always be either "ip:non-shard-aware-port" or "ip:shard-aware-port"? Maybe we could group by the address (if the IPs are the same we could only print ports) and report how many connections are there to each address?

File: scylla/src/transport/connection_pool.rs [hunk 3]
```diff
@@ -80,6 +90,43 @@ enum PoolConnections {
     },
 }
 
+fn format_connection_vector(conns: &[Arc<Connection>]) -> String {
```
Comment (piodul) [inline scylla/src/transport/connection_pool.rs:93]: Could you try converting this and `format_sharded_connection_vector` to a version which uses a `std::fmt::Formatter`, just like the impls for `std::fmt::Display`? It will be more efficient this way, and probably not much more complicated.

While I am aware that this code is for the debugging purposes, I think we should not make them unnecessarily inefficient. Those functions will probably be called frequently if trace level is enabled, so let's make sure that they don't add too much overhead. This is especially important if somebody wants to use it to debug production issues ;)

Comment (psarna) [reply scylla/src/transport/connection_pool.rs:93]: yup, sure, I'll rewrite to std::fmt::Formatter; It's a pity I can't just implement Display for `Vec<Arc<MyType>>`

File: scylla/src/transport/connection_pool.rs (hunk 2 above)
Comment (psarna) [reply scylla/src/transport/connection_pool.rs:98]: Assuming there are no bugs - yeah, sounds reasonable. But here I'm interested in raw data, in case we keep bogus contents in `connections`, so the less processed it is, the better

File: scylla/src/transport/connection_pool.rs (hunk 3 above)
Comment (piodul) [reply scylla/src/transport/connection_pool.rs:93]: If you want to use `Display` then you can introduce a newtype wrapper:

```rust
enum ConnectionVectorWrapper<'a>(&'a Vec<Arc<Connection>>);

impl<'a> std::fmt::Display for ConnectionVectorWrapper<'a> {
    // ...
}

// ...

trace!("My connections: {}", ConnectionVectorWrapper(&conns));
```

File: scylla/src/transport/connection_pool.rs (hunk 2 above)
Comment (piodul) [reply scylla/src/transport/connection_pool.rs:98]: Ah, now I see that this function prints per-shard connections. Grouping by addresses won't reduce verbosity for Scylla connections unless somebody has more than one connection per shard, which sounds rare. Maybe it will help Cassandra users because they are treated as if there was only one shard, but I won't insist too much on implementing it.

**Comment by psarna** [comment]
I'll keep pushing work-in-progress here, I'll let you know when I finish v2 and make it reviewable again

**Comment by psarna** [comment]
Pushed v2 with the review comments addressed. Sample from new logs (they look properly spammy now):
```
Feb 21 12:49:40.337 DEBUG scylla::transport::cluster: Requesting topology refresh
Feb 21 12:49:40.336 DEBUG scylla::transport::connection_pool: [127.0.0.1] Started asynchronous pool worker
Feb 21 12:49:40.338 TRACE scylla::transport::connection_pool: [127.0.0.1] Will open the first connection to the node
Feb 21 12:49:40.338 TRACE scylla::transport::connection_pool: Pool state: 0: 
Feb 21 12:49:40.339 TRACE mio::poll: registering event source with poller: token=Token(1), interests=READABLE | WRITABLE    
Feb 21 12:49:40.339 TRACE scylla::transport::connection: Sending 1 requests; 9 bytes
Feb 21 12:49:40.340 TRACE scylla::transport::connection: Sending 1 requests; 64 bytes
Feb 21 12:49:40.341 TRACE scylla::transport::connection: Sending 1 requests; 58 bytes
Feb 21 12:49:40.341 DEBUG scylla::transport::connection_pool: [127.0.0.1] New sharder: Some(Sharder { nr_shards: 4, msb_ignore: 12 }), clearing all connections
Feb 21 12:49:40.341 DEBUG scylla::transport::connection_pool: [127.0.0.1] Updating shard aware port: Some(19042)
Feb 21 12:49:40.341 TRACE scylla::transport::connection_pool: [127.0.0.1] Adding connection 0x7f1c98001c30 to shard 0 pool, now there are 1 for the shard, total 1
Feb 21 12:49:40.342 DEBUG scylla::transport::connection_pool: [127.0.0.1] Pool is full, clearing 0 excess connections
Feb 21 12:49:40.342 TRACE scylla::transport::connection_pool: Pool state: 0: 127.0.0.1:9042; 1: ; 2: ; 3: 
Feb 21 12:49:40.342 TRACE scylla::transport::connection_pool: Selecting random connection
Feb 21 12:49:40.342 TRACE scylla::transport::connection_pool: Available connections="127.0.0.1:9042"
Feb 21 12:49:40.342 TRACE scylla::transport::connection_pool: Found connection for shard=0
Feb 21 12:49:40.342 TRACE scylla::transport::connection: Sending 3 requests; 247 bytes
Feb 21 12:49:40.344 TRACE scylla::transport::connection: Sending 1 requests; 80 bytes
Feb 21 12:49:40.347 TRACE scylla::transport::connection: Sending 1 requests; 116 bytes
Feb 21 12:49:40.362 TRACE scylla::transport::connection: Sending 1 requests; 100 bytes
Feb 21 12:49:40.366 TRACE scylla::transport::connection: Sending 1 requests; 104 bytes

(...)

Feb 21 12:47:39.157 TRACE Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}: scylla::transport::connection_pool: Selecting connection for token=0
Feb 21 12:47:39.157 TRACE Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}: scylla::transport::connection_pool: Selecting connection for shard=0
Feb 21 12:47:39.157 TRACE Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}: scylla::transport::connection_pool: Available connections="127.0.0.1:19042"
Feb 21 12:47:39.157 TRACE Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}: scylla::transport::connection_pool: Found connection for shard=0
Feb 21 12:47:39.157 TRACE Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}: scylla::transport::session: Sending connection="127.0.0.1:19042"
Feb 21 12:47:39.157 TRACE scylla::transport::connection: Sending 1 requests; 171 bytes
Paging state from the third prepared statement execution: None (2 rows)
Ok.

```

**Comment by psarna** [comment]
I'll add more logs for retry policies and load balancing policies, so please hodl

**Comment by psarna** [comment]
k, done

**Comment by ultrabug** [comment]
> Found connection for shard=0

Would it be useful to know which got selected?

> Adding connection 0x7f1c98001c30 to shard 0 pool, now there are 1 for the shard, total 1

This one looks useful as a DEBUG to me

> Pool is full, clearing 0 excess connections

Is the term "full" appropriate? From a user perspective it's a negative term, maybe "complete" or "fully populated" would be better?

> Sending connection="127.0.0.1:19042"

Sending what?

**Comment by psarna** [comment]
> > Found connection for shard=0
> 
> Would it be useful to know which got selected?

Yeah, definitely - it will be unconditionally printed later - that's the "Sending" line you mention below.

> 
> > Adding connection 0x7f1c98001c30 to shard 0 pool, now there are 1 for the shard, total 1
> 
> This one looks useful as a DEBUG to me
> 
> > Pool is full, clearing 0 excess connections
> 
> Is the term "full" appropriate? From a user perspective it's a negative term, maybe "complete" or "fully populated" would be better?

Now, these two above are preexisting - we can definitely discuss if they belong to a higher logging level, but that fix would go to a separate pull request. And @piodul what do you think about the suggestion about `full`? IIRC it's your code, so you might have an opinion to share.

> 
> > Sending connection="127.0.0.1:19042"
> 
> Sending what?

That's the best part - a request recognizable by its contents or prepared id! `tracing` is amazing enough to provide scopes (or "spans" in their nomenclature. Each request creates its own span, so this information does not have to be duplicated in the code. The request id is right there in the prefix of this line:

> Feb 21 12:47:39.157 TRACE **Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}**: scylla::transport::session: Sending connection="127.0.0.1:19042"

**Comment by piodul** [comment]
> > > Adding connection 0x7f1c98001c30 to shard 0 pool, now there are 1 for the shard, total 1
> > 
> > 
> > This one looks useful as a DEBUG to me
> > > Pool is full, clearing 0 excess connections
> > 
> > 
> > Is the term "full" appropriate? From a user perspective it's a negative term, maybe "complete" or "fully populated" would be better?
> 
> Now, these two above are preexisting - we can definitely discuss if they belong to a higher logging level, but that fix would go to a separate pull request. And @piodul what do you think about the suggestion about `full`? IIRC it's your code, so you might have an opinion to share.

"Fully populated" sounds good to me.

> 
> > > Sending connection="127.0.0.1:19042"
> > 
> > 
> > Sending what?
> 
> That's the best part - a request recognizable by its contents or prepared id! `tracing` is amazing enough to provide scopes (or "spans" in their nomenclature). Each request creates its own span, so this information does not have to be duplicated in the code. The request id is right there in the prefix of this line:
> 
> > Feb 21 12:47:39.157 TRACE **Request{prepared_id="EFA7195FDCB9AC9E70D7CB5E0FAFE700"}**: scylla::transport::session: Sending connection="127.0.0.1:19042"

To be fair, request executions are not differentiable by prepared_id. It's common to send multiple requests with the same prepared_id concurrently.

At first I wanted to suggest adding an atomic counter to the Session so that each execution has a unique ID, but it looks like spans are supposed to have unique IDs already. You can configure `tracing-subscriber` to print out those IDs, however it's only available for the "json" formatting, but not "pretty" (the latter is used by @psarna in his examples): [link](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/format/struct.Format.html#method.with_current_span).

File: scylla/src/transport/connection_pool.rs [hunk 4]
```diff
@@ -80,6 +90,61 @@ enum PoolConnections {
...
+        write!(
+            f,
+            "{}",
+            self.0
+                .iter()
+                .map(|conn| conn.get_connect_address().to_string())
+                .collect::<Vec<String>>()
+                .join(",")
+        )
+    }
```
Comment (piodul) [inline scylla/src/transport/connection_pool.rs:105]: I think you can reduce the number of allocations in your `impl Display`s. Take a look at the docs for `Formatter` ([link](https://doc.rust-lang.org/std/fmt/struct.Formatter.html)), especially the `debug_xyz` family of methods. You can use them to print lists, sets, maps, etc.

For example, this one could be rewritten as:

```rust
f.debug_list()
    .entries(self.0.iter().map(|conn| conn.get_connect_address()))
    .finish()
```

This will change the formatting a bit, there will be additional brackets and `Debug` trait will be used to print IPs. If that's not acceptable, we can stay with the current formatting or you could open-code a for loop and use `write!(f, "{}", conn.get_connect_address())` to print the addresses (taking care of separator commas ofc).

This could be done for other impls, too.

Comment (psarna) [reply scylla/src/transport/connection_pool.rs:105]: Yeah, I was tempted to do that, but I thought that maybe it's considered bad taste to use debug formatting in order to implement non-debug string formatting. But you're right that it makes the implementation considerably simpler, so I'll do that, thx.

Comment (psarna) [reply scylla/src/transport/connection_pool.rs]: I went a little further and just implemented `Debug` for these wrappers - their output is pretty much only used for debugging so it fits the purpose, and now I don't have a moral hangover of implementing `Display` by relying on `Debug`.

File: scylla/src/transport/load_balancing/round_robin.rs [hunk 5]
```diff
@@ -34,10 +34,17 @@ impl LoadBalancingPolicy for RoundRobinPolicy {
...
         let rotation = super::compute_rotation(index, nodes_count);
         let rotated_nodes = super::slice_rotated_left(&cluster.all_nodes, rotation).cloned();
+        trace!(
+            nodes = rotated_nodes
+                .clone()
+                .map(|node| node.address.to_string())
+                .collect::<Vec<String>>()
+                .join(",")
+                .as_str()
+        );
```
Comment (piodul) [inline scylla/src/transport/load_balancing/round_robin.rs:47]: Might be good to include some information that this is related to a round robin policy (in DC aware policy you already print "DC Aware").

File: scylla/src/transport/load_balancing/round_robin.rs [hunk 6]
```diff
@@ -34,10 +34,17 @@ impl LoadBalancingPolicy for RoundRobinPolicy {
         cluster: &'a ClusterData,
     ) -> Box<dyn Iterator<Item = Arc<Node>> + Send + Sync + 'a> {
         let index = self.index.fetch_add(1, ORDER_TYPE);
-
```
Comment (piodul) [inline scylla/src/transport/load_balancing/round_robin.rs:37]: Unnecessary whitespace change?

**Review by piodul — CHANGES_REQUESTED** [review]
There is one more place that I think should be instrumented - speculative execution. Each concurrent execution should be placed in a separate span so that we can differentiate and deinterleave the events.

**Comment by psarna** [comment]
comments addressed (I hope I didn't miss any)

**Comment by psarna** [comment]
oh wait, I did miss one - speculative execution. Coming soon

File: scylla/src/transport/session.rs [hunk 7]
```diff
@@ -1103,10 +1118,20 @@ impl Session {
         let mut last_error: Option<QueryError> = None;
 
         'nodes_in_plan: for node in query_plan {
+            let span = trace_span!("Executing query", node = node.address.to_string().as_str());
```
Comment (piodul) [inline scylla/src/transport/session.rs:1121]: In this function you set the parent span for all `trace!` messages, but what about messages in other functions, called from this one?

Comment (piodul) [reply scylla/src/transport/session.rs:1121]: Ah, you wrapped futures returned by them with `.instrument()`.

File: scylla/src/transport/session.rs [hunk 8]
```diff
@@ -1140,7 +1177,9 @@ impl Session {
                     ),
                 };
 
-                match retry_session.decide_should_retry(query_info) {
+                let retry_decision = retry_session.decide_should_retry(query_info);
+                trace!(parent: &span, "Retry decision: {:?}", retry_decision);
```
Comment (piodul) [inline scylla/src/transport/session.rs:1181]: Nit: `trace!(parent: &span, retry_decision=format!("{:?}", retry_decision))`

File: scylla/src/transport/session.rs (hunk 7 above)
Comment (psarna) [reply scylla/src/transport/session.rs:1121]: Yup. I could still potentially miss local synchronous helper functions that way, but there aren't any that I'm aware of. And unfortunately this instrumentation gets quite ugly because we're using nested named loops here, so I can't just wrap everything in a single instrumentation.

File: scylla/src/transport/session.rs (hunk 8 above)
Comment (psarna) [reply scylla/src/transport/session.rs]: done

File: scylla/src/transport/connection_pool.rs [hunk 9]
```diff
@@ -159,19 +215,26 @@ impl NodeConnectionPool {
         // Try getting the desired connection
         if let Some(conn) = Self::choose_random_connection_from_slice(&shard_conns[shard as usize])
         {
+            trace!(shard = shard, "Found connection for");
```
[the remark this answers is not in the document]
Comment (psarna) [reply scylla/src/transport/connection_pool.rs]: done

File: scylla/src/transport/session.rs (hunk 7 above)
Comment (piodul) [reply scylla/src/transport/session.rs:1121]: I think you can try to move out the `'same_node_retries` loop into an async block or a separate function and make it return `Result<RetryDecision>`. The `match retry_decision` block would be just outside the span.

As for the synchronous functions, we have `retry_session.decide_should_retry` which is synchronous and may run custom user logic, which in turn can print trace messages, too.

Comment (piodul) [reply scylla/src/transport/session.rs:1121]: Hmm, now I think that `Result<RetryDecision>` will be insufficient because we won't be able to return on success... So maybe something like `Result<std::ops::ControlFlow<Result<ResT, QueryError>, RetryDecision>` (ugh, that's a mouthful, but the type inference should make it unnecessary to type it out)?

Comment (psarna) [reply scylla/src/transport/session.rs:1121]: let's move it to a follow-up, I'd rather not mess with the flow in this particular patch (in order to make regressions less likely)

**Review by piodul — APPROVED** [review]
